src/main/java/org/apache/commons/crypto/stream/CryptoOutputStream.java

```diff
@@ -359,12 +379,32 @@ protected void encrypt() throws IOException {
         inBuffer.clear();
         outBuffer.flip();
 
-        // write to output
         while (outBuffer.hasRemaining()) {
             output.write(outBuffer);
         }
     }
 
+    /**
+     * Does the encryption in a non-blocking mode, input is {@link #inBuffer} and output is
+     * {@link #outBuffer}.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    protected void encryptNonBlocking() throws IOException {
```

Comment:
I think instead of this method you should instead remove the following code from the existing `encrypt()`:

```
        // write to output
        while (outBuffer.hasRemaining()) {
            output.write(outBuffer);
        }
```

And move it to the existing blocking call sites (or maybe some helper method).